### ISSUE TITLE
Allow `vim.lsp.util.open_floating_preview` to recycle the floating preview's buffer

### ISSUE THREAD

**Issue by ddickstein** [neovim/neovim#29663]
### Problem

`vim.lsp.util.open_floating_preview` creates a new buffer each time it's called, which can drive up the buffer numbers pretty aggressively, esp. for diagnostics, where the window and buffer tend to be pretty transient. Can some logic be added to reuse an existing buffer, or otherwise make this configurable (e.g., allowing an existing buffer to be passed through that will be overwritten instead of a fresh one being created)?

https://github.com/neovim/neovim/blob/45b7a2c50335e1943a36715101e40eda5a1423f3/runtime/lua/vim/lsp/util.lua#L1606

### Expected behavior

.

**Comment by asmodeus812** [comment]
Well 50 lines down you have this - `vim.bo[floating_bufnr].bufhidden = 'wipe'`. Which means that the buffer will be cleaned up immediately after it goes hidden.

**Comment by ddickstein** [comment]
Correct, but the buffer ids will still tick upward. I don't like buffer ids in the thousands.

**Comment by ddickstein** [comment]
Aesthetically when I'm viewing the buffer list I don't like the really large prefixes. If I decide to manually delete buffers by number, I'd like those numbers to be smaller.

**Comment by asmodeus812** [comment]
I suppose that since you tell a buffer is a 'scratch' type through the api at the moment of creation, and those are usually the ones used for utility purposes, such as the one presented in this issue, nvim could internally treat these buffers a bit different, and assign those buffers different id/seed/starting id compared to regular non-scratch buffers.

**Comment by ddickstein** [comment]
I think that would be a lot more complicated than my proposal of letting
you pass through a buffer to use. Also you can toggle whether a buffer is a
scratch buffer through [these settings](
https://neovim.io/doc/user/windows.html#scratch-buffer)

**Comment by asmodeus812** [comment]
I was thinking a general solution, to a problem, that is a pretty common pattern across many plugins and builtin functionalities. But yeah if you wish this solved only for floating windows you can either pass it in, or have the implementation reuse a buffer.

**Comment by ddickstein** [comment]
I think a general solution to this would be pretty hard, and it's not clear what it would look like. There's a lot of value in being able to uniquely identify windows and buffers for a vim session, so without surrounding context that lets you know you can confidently recycle a buffer doing so would lead to bugs. The signal of whether a buffer is a scratch buffer wouldn't work for this for the reason I gave. Even if you were to do something like enable the reuse of buffer ids of buffers that are wiped out (not just deleted), you still could introduce bugs where plugins store buffer-specific state outside of buffer variables that they would begin applying to the new buffer that claims the id.

**Comment by justinmk** [comment]
> Even if you were to do something like enable the reuse of buffer ids of buffers that are wiped out (not just deleted), you still could introduce bugs where plugins store buffer-specific state

Yeah that's a risk.

There is precedent for reusing a scratch buffer to avoid excessive buffer id "pollution": that's how `inccommand=split` works internally. 

However, it should also be emphasized that buffer ids in general must not be thought of as precious. Buffers are the fundamental editor data structure, and it must be expected and encouraged for plugins to be able to use them as throwaway objects, just like strings, dicts, etc.

# Proposal

AFAICT, in this case it seems to make sense for `open_floating_preview` to keep a circular queue (`vim.ringbuf()`) of scratch buffers, instead of calling `nvim_create_buf` every single time. Whenever a `open_floating_preview` is closed, the buffer is added back to the list of scratch buffers.

If this doesn't add much code, it seems acceptable. It would be even simpler if `open_floating_preview` isn't intended to show more than one float at a time.

**Comment by ddickstein** [comment]
Re: buffer ids not being precious, I think there are some things that just become less ergonomic as the ids climb. For example, buffer commands such as `:sb` and `:bw` when an ID is passed. Though I'm not claiming that a workflow that is heavily reliant on buffer ids even being exposed is a good idea.

Perhaps related - what is the max possible value for a buffer, window, and tab page id? If it's int32 or uint32, what happens when the id reaches its max value? If buffers and windows are seen as throwaway objects that are discarded and recreated rapidly I could see these limits being reached in a long-lived editor session.

**Comment by justinmk** [comment]
> what happens when the id reaches its max value?

No one knows. At int32, if a plugin creates 1 buffer/s, that's 2 billion seconds = ~60 years :)

**Comment by ddickstein** [comment]
Sure, but I think it's a bit more likely to be something like N<10 buffers at 100ms (say, per keystroke). Maybe N = 1, but I'm not sure how much we should rely on that -- you've got completions, diagnostics, and you could imagine people having a few more in more custom configs. If we cut by 2 orders of magnitude, it's still unlikely to be hit, but it's not as comfortable as an int64. Sorry for derailing the discussion though -- your proposal above sounds good to me.

**Comment by mfussenegger** [comment]
Adding some so far not mentioned aspects here:

- `open_floating_preview` is a candidate for removal with a replacement that's not in `lsp.util`: https://github.com/neovim/neovim/issues/25272 and https://github.com/neovim/neovim/issues/25514 (My preference would be to do that first, and basically block `open_floating_preview` for any extension whatsoever)
- There'd be the previewwindow feature for cases where you want to have one unique preview. We could consider using that instead in a few places. There is also a PR to extend it to support float: https://github.com/neovim/neovim/pull/26799
- `open_floating_preview` is currently used for different features - hover, diagnostics, signature_help. If we recycle buffers we need to make sure that any state set from one feature is always reset. Should be doable, just to have it mentioned for the complexity assessment. And I don't know if this wouldn't be in some ways breaking if people relied on certain autocmd's. Or plugins could've cached the bufnr and relied on it becoming invalid if its gone.
- I have in mind to change hover & signature help to support being moved from float to non-float, and if done, no longer auto-close on move to allow users to keep it open to be able to keep reading the docs while they're typing/moving around. Shouldn't conflict with any buffer recycling idea - but just to have it mentioned.
